README.md
# griffith
Griffith Movie Collection Manager


ChangeLog
# $Id$ vim: ts=4

Griffith ChangeLog
------------------
(c) 2005-2014  Vasco Nunes, Piotr Ożarowski


2014-03-13  Michael Jahn
	* [#1290146] imdb searches
	* [#1271874] "weitere Titel" werden in Originaltitel übernommen

2013-12-16  Michael Jahn
	* better handling of temporary poster files
	
2013-12-09  Michael Jahn
	* IMDb plugin updated

2013-12-03  Michael Jahn
	* IMDb plugin updated

2013-11-17  Michael Jahn
	* [#1251923] poster not loaded from CSFD

2013-11-12  Michael Jahn
	* [#1250013] Allocine plugin do not populate the fields

2013-10-17  Michael Jahn
	* IMDb plugin updated
	  [#1235747] IMDB fields not populating
	  [#1239178] IMDB fails to populate some fields
	  [#1236093] IMDB failure starting around 25 Sept 2013
	  [#1240477] IMDB not Downloading All Fields in 0.13 for Linux

2013-10-04  Michael Jahn
	* FilmeAffinity movie plugin updated
	  [#1210828] FilmAffinity
	  [#1216572] [FilmAffinity plugin] doesn't show the list of movies

2013-09-27  Michael Jahn
	* FilmeVonAZ movie plugin updated

2013-06-26  Michael Jahn
	* [1194709] DVDEmpire plugin stopped returning results

2013-06-24  Michael Jahn
	* [1174569] FilmAffinity plugin does not work since today

2013-06-01  Michael Jahn
	* PDF export plugin updated
	
2013-05-24  Michael Jahn
	* Allocine movie plugin updated

2013-04-02  Michael Jahn
	* [1039068] Return the list of films sorted; added sorting by title for result lists

2013-02-21  Michael Jahn
	* IMDb and Cinemovies movies plugin updated

2013-01-30  Michael Jahn
	* FilmeVonAZ and OFDb movie plugins updated

2013-01-28  Michael Jahn
	* Kino.de and Cineteka movie plugins updated

2012-12-29  Michael Jahn
	* remove movie plugin FilmDb (site doesn't exists anymore)

2012-12-16  Michael Jahn
	* fixed backup for databases other than sqlite

2012-12-06  Michael Jahn
	* fixed IMDb plugins for direct search results

2012-11-29  Michael Jahn
	* [1084608] IMDB does not work from November 28, 2012; fixed IMDb plugins

2012-11-20  Michael Jahn
	* fixed Allocine plugin [663709]
	
2012-08-30  Michael Jahn
	* [1043263] importing GriffithXML fails: posters are not added

2012-08-29  Michael Jahn
	* [1037249] typo in tellico spelling
	
2012-07-03  Michael Jahn
	* DVD Palace movie plugin updated
	
2012-06-09  Michael Jahn
	* DVD Palace movie plugin updated

2012-06-07  Michael Jahn
	* [#1010073] The MoviePosterDB seems to have quit working properly.

2012-04-10  Vasco Nunes
	* [#669431] Spell checker doesn't work

2012-04-02  Michael Jahn
	* fixed Allocine plugin (V3 API) (thanks to CinéphOli)

2012-01-29  Michael Jahn
	* [#923416] Connection failed. (Amazon plugin)
	* [#913299] importing covers gives list index out of range
	* [#913283] importer: 'poster' overwritten by 'image'

2012-01-10  Piotr Ożarowski
	* Filmweb movie plugin updated (thanks to mariusz2806)

2012-01-08  Piotr Ożarowski
	* Tellico import plugin added (thanks to Elan Ruusamäe)

2012-01-07  Ivo Nunes
	* Added a plugin for the portuguese website www.dvdpt.com

2012-01-02  Michael Jahn
	* [#909513] code errors in import plugins

2011-12-27  Michael Jahn
	* better error handling for PDF export plugin

2011-12-26  Piotr Ożarowski
	* Restoring database from backup fixed

---- 0.13 ----

2011-12-18  Piotr Ożarowski
	* Filmweb movie plugin updated (thanks to mariusz2806)

2011-12-02  Michael Jahn
	* [#888347] FilmAffinity plugin doesn't work
	* [#897774] removal of the first added subtitle format is not possible

2011-12-01  Michael Jahn
	* [#885914] Amazon Plugin in Griffith doesn't work

2011-09-14  Michael Jahn
	* [#815421] add ipython 0.11 compatibilty (thanks to Julian Taylor)

2011-09-12  Michael Jahn
	* added limited urllib2 support

2011-09-04  Piotr Ożarowski
	* AniDB movie plugin rewritten (uses HTTP XML API now) - needs lxml module

2011-09-01  Michael Jahn
	* updated Allocine plugin, switch to JSON api
	* added database maintenance extension

2011-08-30  Michael Jahn
	* updated plugins: AllRovi, Cinemovies, Cineteka, CSFD, IMDb, Zelluloid

2011-08-23  Michael Jahn
	* added buttons to remove and add a poster within the movie edit dialog
	* fixed bug in extension "remover", it doesn't work if a criteria is required
	  (f.e. adv filter with "required tag" did only remove the tags not the movie entries)
	* extension remover now respects the simple filter condition

2011-08-20  Michael Jahn
	* updated KinoDe plugin
	* new icons for loading and removing posters (toolbar and menu are now the same icons)

2011-08-10  Michael Jahn
	* added new icons for the extensions (thanks to CinéphOli)

2011-07-18  Michael Jahn
	* configurable default value for field "seen it"
	* ability to switch the behaviour of the live search that it queries only after
	  enter is pressed (switchable via griffith.cfg only); good for bigger databases
	* [#811685] IMDB movie ratings not updating when adding new movie

2011-06-22  Filippo Valsorda
	* Fixed and completed PluginMovieMyMoviesIt (hidden bug)
	* Added gutils.rtrim (think at string.rfind)

2011-06-20  Filippo Valsorda
	* Refresh IMDB-it plugin

2011-06-20  Michael Jahn
	* [#570053] Region selection shows DVD regions when media is not DVD.
	  (added Bluray regions)

2011-06-17  Michael Jahn
	* [#725810] CSFD plugin is not working

2011-06-16  Michael Jahn
	* [#721033] Wrong behaviour

2011-06-13  Michael Jahn
	* [#789246] movies.studio field length too short in postgres
	  added a LengthValidator which truncates text fields

2011-06-01  Ivo Nunes
	* fixed crash when you try fetching a movie's information without internet connection

2011-05-23  Michael Jahn
	* updated MyMoviesIt plugin (full cast, screenplay, cameraman, director)
	* [#721036] Problem to add a movie

2011-05-18  Michael Jahn
	* updated CineMovies plugin (thanks to Olivier BOURSIN)
	* small fixes for Allocine and AllRovi plugin
	* added quick search for any field in the main window (thanks to Daniel Fetchinson)

2011-05-12  Michael Jahn
	* added "virtual" resolutions (4:3 Fullscreen, Widescreen, Anamorphic Widescreen)


lib/config.py
# -*- coding: UTF-8 -*-

__revision__ = '$Id$'

# Copyright © 2005-2011 Vasco Nunes, Piotr Ożarowski
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA

# You may use and distribute this software under the terms of the
# GNU General Public License, version 2 or later

import os
import ConfigParser
import logging
log = logging.getLogger("Griffith")


class Config(object):
    subst = {'True': True, 'False': False, 'None': None}

    def __init__(self, file):
        self._file = file
        if not self.load():
            self.make_defaults()

        #self.set_hooks = []

    def get(self, option, default=None, section='main'):
        if not isinstance(option, str):
            option = str(option)
        if not self._cfg.has_option(section, option):
            return default
        else:
            value = self._cfg.get(section, option, False)
            value = value.replace('%%', '%')
            value = Config.subst.get(value, value) # replace 'True' etc.
            return value

    def set(self, option, value, section='main'):
        if not isinstance(option, basestring):
            option = str(option)
        if not isinstance(value, basestring):
            value = str(value)
        if not self._cfg.has_section(section):
            self._cfg.add_section(section)
        value = value.replace('%', '%%')
        self._cfg.set(section, option, value)

    def has_key(self, key, section='main'):
        return self._cfg.has_option(section, key)

    def __setitem__(self, k, v):
        self.set(k, v, section='main')

    def __getitem__(self, key):
        return self.get('main', key)

    def remove_option(self, option, section='main'):
        return self._cfg.remove_option(section, option)

    def keys(self, section='main'):
        return [i[0] for i in self._cfg.items(section)]

    def values(self, section='main'):
        return [i[1] for i in self._cfg.items(section)]

    def items(self, section='main'):
        return self._cfg.items(section)

    def to_dict(self, section='main'):
        d = {}
        for i, j in self._cfg.items(section):
            d[i] = j
        return d

    def save(self, file = None):
        if not file:
            file = self._file
        if not os.path.exists(os.path.split(file)[0]):
            os.makedirs(os.path.split(file)[0])
        self._cfg.write(open(file, 'w'))

    def load(self):
        if os.path.isfile(self._file):
            self._cfg = ConfigParser.SafeConfigParser()
            try:
                self._cfg.read(self._file)
            except Exception, e:
                log.debug(str(e))
                log.error('Cannot parse config file')
                return False
            # check values for number datatypes
            # if current value could not be converted to number datatype
            # the default is set instead
            # some older config files need that correction but also in
            # newer files is it possible, that the user inserts non-convertable
            # values for number fields. The special problem is the string None.
            tmp_value = self.get('color', 0, 'defaults')
            try:
                int(tmp_value)
            except:
                self.set('color', 0, 'defaults')
            tmp_value = self.get('condition', 0, 'defaults')
            try:
                int(tmp_value)
            except:
                self.set('condition', 0, 'defaults')
            tmp_value = self.get('layers', 0, 'defaults')
            try:
                int(tmp_value)
            except:
                self.set('layers', 0, 'defaults')
            tmp_value = self.get('media', 0, 'defaults')
            try:
                int(tmp_value)
            except:
                self.set('media', 0, 'defaults')
            tmp_value = self.get('region', 0, 'defaults')
            try:
                int(tmp_value)
            except:
                self.set('region', 0, 'defaults')
            tmp_value = self.get('vcodec', 0, 'defaults')
            try:
                int(tmp_value)
            except:
                self.set('vcodec', 0, 'defaults')
            return True
        else:
            return False

    def make_defaults(self):
        self._cfg = ConfigParser.SafeConfigParser()
        self._cfg.read(self._file)
        self._cfg.add_section('database')
        self._cfg.add_section('defaults')
        self._cfg.add_section('window')
        self._cfg.add_section('mainlist')
        self._cfg.add_section('mail')
        self._cfg.add_section('spell')
        self._cfg.add_section('add')
        self._cfg.add_section('main')
        self._cfg.set('add', 's_cast', 'True')
        self._cfg.set('add', 's_classification', 'True')
        self._cfg.set('add', 's_country', 'True')
        self._cfg.set('add', 's_director', 'True')
        self._cfg.set('add', 's_genre', 'True')
        self._cfg.set('add', 's_image', 'True')
        self._cfg.set('add', 's_notes', 'True')
        self._cfg.set('add', 's_o_site', 'True')
        self._cfg.set('add', 's_o_title', 'True')
        self._cfg.set('add', 's_plot', 'True')
        self._cfg.set('add', 's_rating', 'True')
        self._cfg.set('add', 's_runtime', 'True')
        self._cfg.set('add', 's_site', 'True')
        self._cfg.set('add', 's_studio', 'True')
        self._cfg.set('add', 's_title', 'True')
        self._cfg.set('add', 's_trailer', 'True')
        self._cfg.set('add', 's_year', 'True')
        self._cfg.set('add', 's_screenplay', 'True')
        self._cfg.set('add', 's_cameraman', 'True')
        self._cfg.set('add', 's_resolution', 'True')
        self._cfg.set('add', 's_barcode', 'True')
        self._cfg.set('database', 'file', 'griffith.db')
        self._cfg.set('database', 'host', 'localhost')
        self._cfg.set('database', 'name', 'griffith')
        self._cfg.set('database', 'passwd', 'gRiFiTh')
        self._cfg.set('database', 'port', '5432')
        self._cfg.set('database', 'type', 'sqlite')
        self._cfg.set('database', 'user', 'griffith')
        self._cfg.set('defaults', 'color', '0')
        self._cfg.set('defaults', 'condition', '0')
        self._cfg.set('defaults', 'layers', '0')
        self._cfg.set('defaults', 'media', '0')
        self._cfg.set('defaults', 'region', '0')
        self._cfg.set('defaults', 'vcodec', '0')
        self._cfg.set('mail', 'email', 'griffith')
        self._cfg.set('mail', 'password', '')
        self._cfg.set('mail', 'smtp_server', 'localhost')
        self._cfg.set('mail', 'use_auth', 'False')
        self._cfg.set('mail', 'username', '')
        self._cfg.set('mail', 'mail_smtp_port', '25')
        self._cfg.set('mail', 'mail_use_tls', 'False')
        self._cfg.set('main', 'default_movie_plugin', 'IMDB')
        self._cfg.set('main', 'font', '')
        self._cfg.set('main', 'font_size', '18')
        if os.name == 'nt' or os.name.startswith('win'):  # win32, win64
            self._cfg.set('main', 'pdf_reader', '')
        else:
            self._cfg.set('main', 'pdf_reader', 'xpdf')
        self._cfg.set('main', 'posters', 'posters')
        self._cfg.set('main', 'rating_image', '0')  # 0 = meter; 1 = stars
        self._cfg.set('mainlist', 'director', 'True')
        self._cfg.set('mainlist', 'image', 'True')
        self._cfg.set('mainlist', 'limit', '0')  # limit search results to x items (0 -> no limits)
        self._cfg.set('mainlist', 'number', 'True')
        self._cfg.set('mainlist', 'otitle', 'True')
        self._cfg.set('mainlist', 'sortby', 'number')
        self._cfg.set('mainlist', 'sortby_reverse', 'False')
        self._cfg.set('mainlist', 'title', 'True')
        self._cfg.set('spell', 'gtkspell', 'True')
        self._cfg.set('spell', 'lang', 'en')
        self._cfg.set('spell', 'notes', 'True')
        self._cfg.set('spell', 'plot', 'True')
        self._cfg.set('window', 'height', '700')
        self._cfg.set('window', 'left', 'None')
        self._cfg.set('window', 'top', 'None')
        self._cfg.set('window', 'view_toolbar', 'True')
        self._cfg.set('window', 'view_ext_toolbar', 'False')
        self._cfg.set('window', 'width', '500')
        self.save()
